### DISCUSSION TITLE
How to optimize performance when rendering large amount of objects ?

### DISCUSSION THREAD

**Discussion by ghost** [fabricjs/fabric.js#7506] [Q&A] [answered]
I've requirement to render a large amount of SVG's as you can see in the screenshot but as the number of objects gets larger and larger the rendering process gets slower. As per the requirement there could be a maximum of 10000 objects and as each of them are SVGs and each SVG have 8 paths therefore there are total 80000 objects. Also each SVG can be coloured independently. So is there any possibility to improve the rendering process somehow ?

![image](https://user-images.githubusercontent.com/37113869/142566567-a989b6bb-f44a-44ad-aca4-645b87cc5777.png)
![image](https://user-images.githubusercontent.com/37113869/142566699-88582b18-e9f9-4719-a5c5-56640c596737.png)
![image](https://user-images.githubusercontent.com/37113869/154897446-47444a52-6793-4e8d-ac8c-954a93042a7f.png)

**Comment by ShaMan123** [comment, marked as answer]
If all objects are the same this should work fine in some cases.
Take for example `SprayBrush` which is made of thousands of `fabric.Circle`.
It seems you need to look at `objectCaching`.
And if I were you I would parse the svg once and clone the fabric object again and again (and try to minimize the path parsing as much as possible).
You should consider subclassing group depending on your needs.

**Reply by ghost** [reply]
@ShaMan123 I'm not sure how `objectCaching` works, but the pattern also fails to render sometimes when I load it for the first time. So I think `objectCaching` would not work.
Also the svgs can be coloured independently (plz see the screenshot in Question), so I think object cloning would also not work.
And I'm not sure about how subclassing group would help in this case, could you please elaborate ?

**Reply by ShaMan123** [reply]
`objectCaching` is a very clever performance optimization. If an object is caching it creates it's own canvas and renders on to that every time it changes. This way when the canvas renders the object simply paints the cache over it, which is a lot more faster than recreating everything each render. It is probably not related to the load failure.
http://fabricjs.com/fabric-object-caching
http://fabricjs.com/svg-caching

If all of these diamonds are the same in shape I would create a custom object (a group) that contains these paths inside it with the ability to change their color.
I would then do the following:
1. create a memoize function for the caches so all these diamonds share the cache canvas reducing memory and computation exponentially. In particular, have the paths create a cache for each different path, and for every change made to them. Then the group will simply render the paths from cache. So if you have 10^5 diamonds but only 50 different paths (in shape and color) if will create 50 cache canvases for the paths. You can continue with this approach to the groups as well but it depends on the variance of combinations you have. If it's little like in the image you've posted then this approach will favor for groups as well. If variance if great I am not sure it will be good for performance to create so many cache canvases. (2) is then critical.
2. Take a look at an experiment I've created #7675. I suggest you do the same. The idea is to group batches of shapes into a group and have it `objectCaching` true. This way it will iterate over it's diamonds only once (if they don't change), render a cache and after that every render cycle it will render the cache regardless of the diamonds. If they do change then the change doesn't trigger a massive rendering. Instead it will be contained to one of the buffering groups. This is an approach for **partial re-render** as @asturur mentioned

**Reply by ghost** [reply]
I tried by creating a diamond with `objectCaching = true` and then used the `diamond.clone()` to create more diamonds but it seems like the `diamond.clone()` method is not using the cached object, it's creating a new one, because it is taking the same amount of time to render those diamonds on the canvas.
If I use the `diamond.cloneAsImage()` then it's using the cached object, and the render time is reduced but the diamonds looks blurry when I try to zoom the canvas.
Any suggestion @ShaMan123 ?

**Reply by ShaMan123** [reply]
blurry is because of retina scaling probably.
clone is exactly like creating an object.
You can configure clone and that might help.
maybe try setting the cache canvas of the clone to the original's before it renders
This is why I suggested the 2 approaches seen above

**Reply by ghost** [reply]
@ShaMan123 can you please guide me through how can I configure the clone to use the original's cache. I checked the fabricJS doc and tried enough but I'm not able to configure the clone.

I tried to configure the clone like this, but it's not working -

```
/* creates a new diamond by grouping some paths and returns that group,
Also the group is created with `objectCaching = true` */
var diamond = createDiamond(); 

canvas.renderAll(); // render the canvas, so that diamond's cache gets created

diamond.clone(function(clone) {
clone.set({
  left: 100,
  _cacheCanvas: diamond. _cacheCanvas // set the cloned diamond cache to the original's
})
canvas.renderAll();
```

**Reply by ShaMan123** [reply]
set up a fiddle or sandbox and I'll try to help.
I think setting `_cacheCanvas` is the right way, maybe you need to override `updateCacheCanvas` and set `dirty`  to false and override `shouldCache` and `isCacheDirty`

**Reply by ShaMan123** [reply]
Current logic prevent an object from caching if it's parent is caching, that's why you must override `shouldCache`

**Reply by ghost** [reply]
@ShaMan123  I have set up a sandbox. Please check it out. The rendering is optimized for the first load, but it takes a lot of time after I change a diamond's colour. There are two data files named `data-small.json` and `data-big.json` for small and big pattern. Please check with both the data files.
Please suggest, how can I optimize the code further.

sandbox link - [https://codesandbox.io/s/frosty-dew-9lb99b?file=/src/App.js](url)

**Reply by ShaMan123** [reply]
It looks good.
I experienced very good performance with the small data file.
With the big data file first rendering is bad, and zooming as well, but color change is lightning fast.
Try implementing another caching mechanism on paths as well (same as you did) in `getKnotTile` (If I got it right).
Make sure you override `shouldCache`, without it an object under a group will not use caching no matter what.
Override `isCacheDirty` as well
Then try to create batching/buffering groups as I mentioned previously with the same override for `shouldCache`

**Reply by ShaMan123** [reply]
changed `cacheClonedNode`, `getKnotTileNode`
https://codesandbox.io/s/gifted-morse-sm9ss4?file=/src/App.js

rendered size is wrong

**Reply by ShaMan123** [reply]
Inspired by what @asturur suggested, you might want to calculate in advance how many diamonds you should render to initially fit the screen with `skipOffscreen` to false and then after rendering has finished render the rest in batches

**Reply by ShaMan123** [reply]
Another good idea to check out is to render a pattern of the colorless diamond on canvas with `repeat` option.
And then add only diamonds that a different to canvas on top of the pattern

**Reply by ShaMan123** [reply]
You can also parse the path data with a util from fabric in advance, that will save some computation effort a well

**Comment by asturur** [comment]
if you draw them one by one is going to be slow.
This is a situation where partial re-renders of the canvas could help, but is an edge case.

Can you optimize the 8 paths in a single one?

**Reply by ghost** [reply]
@asturur could you please explain about the partial re-renders, I'm not sure how this works.
Also as you can see in the question the SVGs can be coloured independently, so I cannot optimize the 8 paths in a single one.

What if I replace fabric with only SVG, I mean if I do not use fabric/canvas and only render SVGs on DOM. Do you think this will help ?

**Reply by asturur** [reply]
If you render as SVG in the dom you have to rebuild the interaction layer.
Yes you probably get some speed, but 80.000 svg are many also for the DOM.
Did you try to to the same with React? how does it perform?

Partial re-renders ( not existing yet ) would redraw just the smallest possible rectangular area of the canvas that has been influenced by your changes.
Is something i wanted to work since long.

Can you explain me how the interaction of the app work?
There is probably a quick use case that can be develop as a test

**Comment by asturur** [comment]
Also in this case, when you know you are at zoom level where objects are all visibile, disabling the offscreen checks save you some times too.
